src/Storages/ObjectStorage/DataLakes/DataLakeConfiguration.h

```diff
@@ -144,15 +150,235 @@ class DataLakeConfiguration : public BaseStorageConfiguration, public std::enabl
+        ObjectStoragePtr object_storage,
+        const Strings & requested_columns,
+        const StorageSnapshotPtr & storage_snapshot,
+        bool supports_subset_of_columns,
+        ContextPtr local_context) override
+    {
+        return getImpl().prepareReadingFromFormat(
+            object_storage,
+            requested_columns,
+            storage_snapshot,
+            supports_subset_of_columns,
+            local_context);
+    }
+
+    std::optional<ColumnsDescription> tryGetTableStructureFromMetadata() const override
+        { return getImpl().tryGetTableStructureFromMetadata(); }
+
+    void update(ObjectStoragePtr object_storage, ContextPtr local_context) override
+        { return getImpl().update(object_storage, local_context); }
+
+    void initialize(
+        ASTs & engine_args,
+        ContextPtr local_context,
+        bool with_table_structure,
+        std::unique_ptr<StorageObjectStorageSettings> settings) override
+    {
+        createDynamicConfiguration(engine_args, local_context);
+        getImpl().initialize(engine_args, local_context, with_table_structure, std::move(settings));
+    }
+
+    ASTPtr createArgsWithAccessData() const override
+    {
+        return getImpl().createArgsWithAccessData();
+    }
+
+protected:
+    void fromNamedCollection(const NamedCollection & collection, ContextPtr context) override
+        { return getImpl().fromNamedCollection(collection, context); }
+    void fromAST(ASTs & args, ContextPtr context, bool with_structure) override
+        { return getImpl().fromAST(args, context, with_structure); }
+
+    /// Find storage_type argument and remove it from args if exists.
+    /// Return storage type.
+    ObjectStorageType extractDynamicStorageType(ASTs & args, ContextPtr context, ASTPtr * type_arg = nullptr) const override
+    {
+        static const auto storage_type_name = "storage_type";
+
+        if (auto named_collection = tryGetNamedCollectionWithOverrides(args, context))
+        {
+            if (named_collection->has(storage_type_name))
+            {
+                return objectStorageTypeFromString(named_collection->get<String>(storage_type_name));
+            }
+        }
+
+        auto type_it = args.end();
+
+        /// S3 by default for backward compatibility
+        /// Iceberg without storage_type == IcebergS3
+        ObjectStorageType type = ObjectStorageType::S3;
+
+        for (auto arg_it = args.begin(); arg_it != args.end(); ++arg_it)
+        {
+            const auto * type_ast_function = (*arg_it)->as<ASTFunction>();
+
+            if (type_ast_function && type_ast_function->name == "equals"
+                && type_ast_function->arguments && type_ast_function->arguments->children.size() == 2)
+            {
+                auto name = type_ast_function->arguments->children[0]->as<ASTIdentifier>();
+
+                if (name && name->name() == storage_type_name)
+                {
+                    if (type_it != args.end())
```

Comment:
I would use an extra boolean variable for clarity: `bool found_storage_type_argument`

Reply:
I used the same way as in `StorageURL::evalArgsAndCollectHeaders` already.
https://github.com/ClickHouse/ClickHouse/blob/master/src/Storages/StorageURL.cpp#L1503
It's a good idea to refactor it later and make a common method to extract key-value arguments, let's make this change in that method.